### ISSUE TITLE
Move to front-end, interactive uploader

### ISSUE THREAD

**Issue by au5ton** [cougargrades/web#36]

**Comment by au5ton** [comment]
Potentially useful:
- https://github.com/mholt/PapaParse
- https://github.com/react-dropzone/react-dropzone
- https://overreacted.io/making-setinterval-declarative-with-react-hooks/ (for progress bar estimations?)

**Comment by au5ton** [comment]
Firestore writes from the front-end can now be checked in Firestore security rules without a document read for every operation.

https://github.com/cougargrades/api/issues/21

**Comment by au5ton** [comment]
https://firebase.google.com/docs/firestore/query-data/listen

**Comment by au5ton** [comment]
https://github.com/dfahlander/Dexie.js

**Comment by au5ton** [comment]
This just occurred to me. Patchfiles are designed in a way that's capable of being asynchronous per phase. Within each "phase", all patchfiles can be done in any order. However, each phase must be done one after another.

If the front-end is to have an interactive uploader, this (effectively) makes Patchfiles more suited to be processed in the front-end instead of in a `patchfile_queue` collection as previously assumed.

This also reduces the amount of "queue tracking" logic necessary to implement because the Patchfiles can quickly out-number the original CSV data in size.

**tl;dr it's easier to code and financially cheaper if Patchfiles are processed client-side**

**Comment by au5ton** [comment]
Following the idea discussed [here](https://github.com/cougargrades/web/issues/36#issuecomment-818471005), I moved the Patchfile execution code into the common module so that the server and front-end are always using the same Patchfile implementation. This means that it's actually really easy now for the front-end to execute Patchfiles.

See changes:
- https://github.com/cougargrades/types/commit/93e84e4ca6bf94c10d987a29cc9c05869b447ec4, available in `>= 0.0.52`
- https://github.com/cougargrades/api/commit/bc4973e7aae18a14ec57fc4e2873b886e7d79062

**Comment by au5ton** [comment]
Papaparse conversion could look like this: 

https://github.com/cougargrades/importer/blob/3e1fe9571e367dd6d7023c80f57a05dcbcd655ff/src/reader.ts#L35-L70

This could be standardized in the types module.

**Comment by au5ton** [comment]
https://github.com/sindresorhus/p-queue

**Comment by au5ton** [comment]
As of 2c80fd740a3ba9348d407bf7ad5560ed2421f4a7, this works for CSV records!! 🎉 

Next:
- Patchfiles
- edge case for upload_queue_backlog:

    > actually, that gives me an idea
    > failed uploads go into the "upload_queue_backlog" collection, i could keep the throttle really high, and at the end do the "backlog" collection of problematic sections and do them at a fixed slower rate
    > everything: 64 concurrent
    > backlog: 4 concurrent

**Comment by au5ton** [comment]
A **complete** working draft of this is in 112da8d0e4adcc7f2e3ca516fc21446de5fffe73, however I have yet to try an entire database upload. I will try soon.

**Comment by au5ton** [comment]
TODO: Estimated completed time and/or speed indicators

perhaps reference: https://github.com/whizsid/react-progress-timer/blob/master/src/ProgressTimer.tsx

**Comment by au5ton** [comment]
Uploader test successful

**Comment by au5ton** [comment]
Certain Patchfile tests failed, but this typo was the culprit: https://github.com/cougargrades/publicdata/commit/3bcdd5b244901530d25481203fc37b355a3a4a59

A full upload will be retried now that this has been fixed. The changes made to resolve https://github.com/cougargrades/api/issues/25 will also be tested.

**Comment by au5ton** [comment]
Notes from yesterdays test:
- All but some patchfiles at the very end succeeded in executing: 
![Screenshot 2021-05-18 021449](https://user-images.githubusercontent.com/4109551/118692190-670dc100-b7cf-11eb-9d9e-a4d4447b009b.png)
- [x] There were some transaction contention issues, so we definitely need to handle the "backlog" more delicately than just adding it back to the queue. It is very possible that backlog re-adds creates a feedback loop. Perhaps, every time something is added to the backlog, the concurrency limit is halved? (No lower than 1 at a time)
- [x] In Groups, we are counting courses_count wrong, so I think we should just remove the property entirely because implementing it is difficult due to the nature of FieldValue.arrayUnion(). This is likely why I didn't include a similar count in Instructors and Courses for their references.
- [x] We need to have Patchfiles which initialize the default Groups, like the core curriculum ones. Currently, they start off empty.
- [x] We also need to assign "categories" to groups. Perhaps renaming these to "tags" would be better. These "tags" can be populated by default as "subject_group" or something in GDR.toGroup() and populated as "core_curriculum" in the core curriculum defaults. Be mindful of "tag" values because they're supposed to be user-friendly strings. Maybe hashtag-like identification, like Discord?
- [x] Maybe we should also consider giving groups an "icon" or "emoji" field so that identifying them in the UI would look more intuitive. We have already experimented with this idea before: https://github.com/cougargrades/web/blob/aa0b8d3f7cf1b76d7c25f5734d047c0a2978027b/src/components/courses/results/IconKey.js

**Comment by au5ton** [comment]
https://stackoverflow.com/questions/55724642/react-useeffect-hook-when-only-one-of-the-effects-deps-changes-but-not-the-oth

**Comment by au5ton** [comment]
I'd say that this is completed.
